contracts/roulette/lib.rs

```diff
@@ -0,0 +1,388 @@
+        },
+        prelude::{format, string::String},
+        reflect::ContractEventBase,
+        storage::{traits::ManualKey, Lazy, Mapping},
+    };
+    use scale::{Decode, Encode};
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum RouletteError {
+        InkEnvError(String),
+        ArithmethicError,
+        BetAmountIsTooSmall,
+        PlayerAlreadyPlacedABet,
+        NoMoreBetsCanBeMade,
+        BettingPeriodNotOver,
+        NativeTransferFailed(String),
+        NotEnoughBalance,
+        CallerIsNotTheHouseOwner,
+    }
+
+    impl From<InkEnvError> for RouletteError {
+        fn from(e: InkEnvError) -> Self {
+            RouletteError::InkEnvError(format!("{e:?}"))
+        }
+    }
+
+    pub type Selector = [u8; 4];
+    pub type Result<T> = core::result::Result<T, RouletteError>;
+    pub type Event = <Roulette as ContractEventBase>::Type;
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct BetPlaced {
+        #[ink(topic)]
+        player: AccountId,
+        #[ink(topic)]
+        bet_type: BetType,
+        amount: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct WheelSpin {
+        winning_number: u8,
+    }
+
+    #[derive(Debug, Encode, Decode, Clone, Copy, PartialEq, Eq)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
+    pub enum BetType {
+        Number(u8),
+        Red,   // even
+        Black, // red
+    }
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
+    pub struct Bet {
+        pub player: AccountId,
+        pub bet_type: BetType,
+        pub amount: Balance,
+    }
+
+    #[derive(Debug)]
+    #[ink::storage_item]
+    pub struct Data {
+        /// represents the contract owner, defaults to the initializer of the contract
+        pub house: AccountId,
+        /// How long does the betting period last? (measured in blocks)
+        pub betting_period_length: BlockNumber,
+        /// When did this betting period start? (measured in blocks)
+        pub betting_period_start: BlockNumber,
+        /// accounting: consecutive bet identifiers
+        pub next_bet_id: u32,
+        /// maximal number of bets that can be made in a round
+        pub maximal_number_of_bets: u8,
+        /// minimal amount of native tokens that can be transferred as part of a bet
+        pub minimal_bet_amount: Balance,
+        /// keeps track of the total potential payouts to make sure all bets can be covered
+        pub potential_payouts: Balance,
+    }
+
+    #[ink(storage)]
+    pub struct Roulette {
+        pub data: Lazy<Data, ManualKey<0x44415441>>,
+        pub bets: Mapping<u32, Bet, ManualKey<0x42455453>>,
+        /// accounting: maps accounts to bets
+        pub account_id_to_bet: Mapping<AccountId, u32>,
```

Comment:
Can a single account place multiple bets? It is common for a single gambler to place their color tiles all over the board in a single spin.

Actually, this is more of a discussion of what should we require students to implement? Or should we let them decide. I think in cases like this we should decide so we can unit test for it.

Reply:
not right no, there is an explicit check. Do we have the requirements for these contracts written down smwhr?

Reply:
> Do we have the requirements for these contracts written down smwhr

Not yet. We just have [my reference impl](https://github.com/Polkadot-Blockchain-Academy/Contract-Writing-Workshop/blob/main/evm/src/Roulette.sol) and now your reference impl.

Part of getting that contract writing workshop ready is making the requirements more clear and providing unit tests for them, and then finally removing the solution so it is just starter code for students to fill in.

Definitely, I didn't clearly state that multiple bets should be supported. And didn't mean any offense or anything negative at all. Just once I saw your design, I realized there were multiple valid options and started the conversation. In this particular case, I kind of prefer multiple bets, but what do you think?

Reply:
no worries, just clarifying. I'll try to implement the comments sometime ~next week , modulo other engagements

Reply:
changed to allow for multiple bets per account